File: apps/model-prompt-guard/src/model/inference.py
"""Dummy prompt injection detection.

This detects common prompt injection patterns using keyword matching.
Will be replaced with actual ML model later.
"""

import re

# Common prompt injection patterns
INJECTION_PATTERNS = [
    r"ignore\s+(previous|all|any)\s+(instructions?|prompts?)",
    r"disregard\s+(previous|all|any)",
    r"forget\s+(everything|previous|all)",
    r"you\s+are\s+now",
    r"new\s+instructions?",
    r"jailbreak",
    r"bypass\s+(safety|filters?|restrictions?)",
    r"reveal\s+(secrets?|hidden|system)",
    r"system\s+prompt",
    r"(\[|\{)\s*(system|SYSTEM)",
    r"pretend\s+(you\s+are|to\s+be)",
    r"act\s+as\s+(if|a)",
    r"roleplay",
    r"DAN\s+mode",
]

COMPILED_PATTERNS = [re.compile(p, re.IGNORECASE) for p in INJECTION_PATTERNS]


def detect_prompt_injection(text: str) -> tuple[bool, float, list[str]]:
    """
    Detect prompt injection attempts in text.
    
    Returns:
        Tuple of (flagged, score, details)
    """
    matches = []
    for pattern in COMPILED_PATTERNS:
        if pattern.search(text):
            matches.append(f"Pattern matched: {pattern.pattern}")
    
    if matches:
        # Score based on number of matches
        score = min(0.5 + (len(matches) * 0.15), 1.0)
        return True, score, matches
    
    return False, 0.0, []

File: apps/model-prompt-guard-go/internal/inference/inference.go
// Package inference provides the prompt injection detection logic.
package inference

import (
	"strings"
)

// Injection keywords to detect
var injectionKeywords = []string{
	"ignore",
	"ignore previous",
	"disregard",
	"forget",
	"bypass",
	"override",
	"reveal",
	"expose",
	"jailbreak",
	"pretend",
	"roleplay",
	"act as",
	"you are now",
	"ignore all",
	"system prompt",
	"instructions",
	"secret",
	"password",
	"confidential",
}

// DetectPromptInjection performs keyword-based prompt injection detection.
// This is a dummy implementation that simulates real ML inference.
//
// Returns:
//   - flagged: true if injection detected
//   - score: confidence score (0.0-1.0)
//   - details: list of reasons
func DetectPromptInjection(text string) (bool, float64, []string) {
	lowerText := strings.ToLower(text)
	var matchedKeywords []string

	for _, keyword := range injectionKeywords {
		if strings.Contains(lowerText, keyword) {
			matchedKeywords = append(matchedKeywords, keyword)
		}
	}

	if len(matchedKeywords) == 0 {
		return false, 0.0, nil
	}

	// Calculate score based on number of matches
	score := float64(len(matchedKeywords)) / float64(len(injectionKeywords))
	if score > 1.0 {
		score = 1.0
	}
	if score < 0.5 {
		score = 0.5 // Minimum confidence for flagged content
	}

	details := []string{"Injection keywords detected: " + strings.Join(matchedKeywords, ", ")}

	return true, score, details
}
